[yongqqqq] iP

yongqqqq opened the pull request:
### Fishman
> "use something that doesn't cost money." 

Fishman helps you to keep track of the tasks you need to do. It's,

- text-based
- easy to learn
- ~paid~ **FREE**

All you need to do is,

1. download it from [here](https://github.com/yongqqqq/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

Features:

- [ ] Manage tasks
- [ ] Manage deadlines
- [ ] Give reminders

---
Below is a _snippet_ of the `main` method:

```java
public class Fishman {
  public static void main(String[] args) throws FishmanException {
          new Fishman().start();
      }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

yongqqqq committed:
Update name of chatbot

yongqqqq committed:
Implement initial skeletal greeting

yongqqqq committed:
Implement command echo and exit function

yongqqqq committed:
Update variables final where suitable

yongqqqq committed:
Add Task class

yongqqqq committed:
Add TaskList class

yongqqqq committed:
Add Command interface

yongqqqq committed:
Add AddCommand class

yongqqqq committed:
Implement command parsing and basic commands

This commit introduces a new Parser class and implements basic commands.

Changes include:
- Add Parser class to read user input and handle Add,"bye" and List
- Implement ExitCommand to end chatbot on "bye" command
- Implement ListCommand to display tasks

These changes improves the structure by separating the logic for
handling user inputs and command classes for various actions. This will
allow for new actions to be easily added in the future.

The AddCommand, ExitCommand and ListCommand provides the basic chatbot
functionality
of adding tasks, exiting the program and displaying the tasklist.

yongqqqq committed:
Update Command interface

- Update Command interface to use TaskList directly instead of generics

yongqqqq committed:
Refactor Fishman class to integrate parsing

This commit refactors the Fishman class to use the new Parse class to
handle user inputs and execution of commands. This improves the overall
structure of the code.

Changes include:
- Integrate Parse for reading of command
- Simplify command execution flow
- Maintain separation of UI, Tasks and Commands

This changes allow for easier extensions in the future.

yongqqqq committed:
Update Ui class with task related display

This commit updates the Ui class to display the newly added
command operations.

Changes include:
- Add displayAddedTask to echo task addition
- Add displayTaskList to show all tasks

This ensures proper integration and separation between the UI and
task operations.

yongqqqq committed:
Add JavaDoc comments to all classes

This will provide better clarity and guidance for others to better
understand the code.

yongqqqq committed:
Add MarkCommand to mark/unmark tasks

This commit introduces the new MarkCommand class which implements
the mark/unmark functionality of fishman bot.

This changes provides the functionality required while allowing
future extensions to the mark feature.

yongqqqq committed:
Add support for Mark/Unmark commands

- Implement MarkCommand to mark tasks as done or not done.
- Updated Parser to recognize "mark" and "unmark" commands.

This improves fishman bot functionality, allowing users to mark tasks
as complete or incomplete.

yongqqqq committed:
Add Deadline, Event and Todo Classes

The classes are implemented to allow for the creation of new tasks
with attributes such as due dates, start and end times as well as tasks
without any specific timing.

This allows for further extensions according to future requirements.

yongqqqq committed:
Update AddCommand, Parser, Task and Ui classes

This commit updates the logic behind the parsing of commands and
tasklist to manage the newly created task types.

This provides a better structure, allowing users to categorise tasks based on their requirements.

yongqqqq committed:
Update test input and expected output

yongqqqq committed:
Updated display in Ui

-Removed additional space from header line.

yongqqqq committed:
Add FishmanException to handle exceptions

-Add FishmanException as base exception class.
-Add InvalidCommandException and EmptyDescriptionException.

yongqqqq committed:
Update FishmanException to handle more exceptions

This commit introduces EmptyListException, MissingArgumentException and
NumberFormatException.

EmptyListException manage cases where a task list is empty and
a mark/unmark command is issued. This ensures that the user receives an
error message when they attempt to mark/unmark tasks in an empty list.

MissingArgumentException checks and ensures that each command and the
respective information required is keyed in by the user. This ensures
that the user receives an error message when there is missing
information.

NumberFormatException manages cases where integer values are expected
but are not inputted by the user. This ensures that the user receives an
error message and can make amends to the command.

yongqqqq committed:
Update Fishman and Parser to throw exceptions

The Parser class will now throw exceptions according to the errors.

The Fishman class has been enhanced to throw an exception if there are
unexpected errors that are not already caught.

This improves error handling by ensuring that the commands/operations
on fishman bot are properly managed.

yongqqqq committed:
Update Ui to handle new exception message

Modified Ui to display error messages from FishmanException.
This will ensure that the user receives accurate feedback on the various
error conditions and how to correct it.

-Add displayError method to handle exception messages.

yongqqqq committed:
Update JavaDocs for multiple classes

Changes include:
- Fishman
- FishmanException
- Parser
- Ui

yongqqqq committed:
Add delete task functionality

This commit includes the creation of DeleteCommand class which is
used to remove tasks from the task list.

yongqqqq committed:
Update Parser,TaskList and Ui to handle deletion

This commit updates Parser and TaskList class to contain logic for
deletion of task from the tasklist given the correct command.
This ensures that fishman bot is able to parse and execute the delete
command correctly.

The Ui class is updated to handle the display of task deletion, allowing
users to receive the appropriate feedback.

yongqqqq committed:
Add IndexOutOfBoundsException and index validation

This commit adds IndexOutOfBoundsException to FishmanException to handle
cases when a provided index is out of bounds. This will display the
appropriate error message to the user.

yongqqqq committed:
Update Deadline, Event, Task, TaskList

yongqqqq committed:
Implement Save functionality

yongqqqq committed:
Revert "Implement Save functionality"

This reverts commit 8e1b1ed1488f8fcf4930a4e398edcd055e5cf005.

yongqqqq committed:
Implement Save functionality

yongqqqq committed:
Add Exception handling for data file

yongqqqq committed:
Merge branch 'branch-Level-7'

yongqqqq committed:
Update Fishman Chatbot to understand datetime

yongqqqq committed:
Merge branch 'branch-Level-8'

yongqqqq committed:
Move classes into suitable Java packages

yongqqqq committed:
Merge branch 'A-Packages'

yongqqqq committed:
Merge branch 'add-gradle-support'

yongqqqq committed:
Update exception handling for user inputs

yongqqqq committed:
Add JUnit tests for Parser

yongqqqq committed:
Add JUnit tests for Storage

yongqqqq committed:
Merge branch 'A-JUnit'

yongqqqq committed:
Add Javadocs for Storage and FishmanException

yongqqqq committed:
Fix Java code style

yongqqqq committed:
Add Find functionality in Fishman bot

yongqqqq committed:
Fix minor codestyle issues

Changes include:
- DeleteCommand
- Ui

yongqqqq committed:
Merge branch 'A-CodingStandard'

yongqqqq committed:
Update Javadoc in FindCommand and Parser

yongqqqq committed:
FindCommand: Remove unused imports

yongqqqq committed:
Fix merge conflict

yongqqqq committed:
Fix Javadoc comments

Changes include:
-DeleteCommand
-ListCommand
-MarkCommand
-FishmanException
-Deadline

yongqqqq committed:
Merge branch 'A-JavaDoc'

yongqqqq committed:
Fix code style

yongqqqq committed:
Merge branch 'A-CheckStyle'

yongqqqq committed:
Add GUI to Fishman bot

amitha2210 commented on src/main/java/fishman/command/MarkCommand.java:
```diff
@@ -0,0 +1,48 @@
+    /** The index of the task in the task list to be marked. */
+    private final int index;
+    /** The boolean to determine whether to mark or unmark the task. */
+    private final boolean isMark;
```
Maybe isMarked would be better than isMark?

amitha2210 commented on src/main/java/fishman/utils/Storage.java:
```diff
@@ -0,0 +1,170 @@
+     * @param task The task to be converted to CSV format.
+     * @return A string representing the task in CSV format.
+     */
+    private String toCsv(Task task) {
```
Good naming here

amitha2210 commented on src/main/java/fishman/utils/Ui.java:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * Displays the welcome message.
+     */
+    public void displayWelcome() {
```
Maybe can name this displayWelcomeMessage instead?

amitha2210 commented on src/main/java/fishman/exception/FishmanException.java:
```diff
@@ -0,0 +1,148 @@
+     *
+     * @param message The detail message.
+     */
+    public FishmanException(String message) {
```
Nice naming of all exception classes, it is clear what each one is for

oolongwater replied:
Exception class naming is really well done here

amitha2210 reviewed (commented):
Overall easy to read and understand naming 👍

oolongwater commented on src/main/java/fishman/Fishman.java:
```diff
@@ -0,0 +1,75 @@
+                ui.displayError("Uh oh, an unexpected error has occured: " + e.getMessage());
+            }
+        }
+        storage.save(tasks);
```
advisable to wrap the storage.save(tasks) in a try-catch block to handle potential IO exceptions and prevent the program from crashing unexpectedly.

oolongwater commented on src/main/java/fishman/utils/Parser.java:
```diff
@@ -0,0 +1,131 @@
+        }
+    }
+
+    private static LocalDateTime parseDateTime(String dateTimeStr) throws FishmanException {
+        try {
+            return LocalDateTime.parse(dateTimeStr, INPUT_FORMATTER);
+        } catch (DateTimeParseException e) {
+            throw new FishmanException.InvalidDateFormatException(dateTimeStr);
+        }
+    }
```
consider adding a Javadoc comment for the private method `parseDateTime` to maintain consistency and provide context for future developers.

oolongwater commented on src/main/java/fishman/utils/Parser.java:
```diff
@@ -0,0 +1,131 @@
+     * @throws FishmanException.IndexOutOfBoundsException if trying to choose a nonexistent item from the list.
+     * @throws FishmanException for unexpected errors during parsing.
+     */
+    public static Command parse(String userInput, TaskList tasks) throws FishmanException {
```
There is some duplication in the code when checking the task list size and parsing indices for `mark`, `unmark`, and `delete`. Consider extracting this logic into a helper method to reduce redundancy.

oolongwater commented on src/main/java/fishman/utils/Parser.java:
The repeated use of command strings like "bye", "list", "mark", etc., could be improved by defining them as constants or using an Enum to represent different command types

oolongwater commented on src/main/java/fishman/utils/Ui.java:
```diff
@@ -0,0 +1,140 @@
+     *
+     * @param message The message header for the error.
+     */
+    public void displayError(String message) {
```
very good naming here

oolongwater reviewed (commented):
The code is well-structured, effectively handling user input parsing and command creation while adhering to most Java conventions. 🥇

yongqqqq committed:
Update error handling for file loading

yongqqqq committed:
Update error handling for file saving

yongqqqq committed:
Update storage unit test

yongqqqq committed:
Update JavaDocs

yongqqqq committed:
Fix checkstyle errors

yongqqqq committed:
Update css and view for gui

yongqqqq committed:
Merge branch 'branch-Level-10'

yongqqqq committed:
Add assertions

Assertions are added to check that certain values are valid
(non-empty or not null).

This ensures that the invalid values are caught, improving
overall robustness.

Let's add null and empty assertions checks to verify
* Deadline, Event, ToDo and TaskList data classes
* Parser and Storage utility classes
* FishmanException exception class
* Fishman main class

yongqqqq committed:
Merge pull request #1 from yongqqqq/branch-A-Assertions

Add assertions

yongqqqq committed:
Refactor Command Classes

Some of the names and comments are unclear or redundant.

Let's refactor them to make it clearer.

yongqqqq committed:
Refactor gui classes

Some of the names and comments are unclear.

Let's refactor them to make it clearer.

yongqqqq committed:
Refactor Task classes

Some of the names and comments are unclear.

Let's refactor them to make it clearer.

yongqqqq committed:
Refactor Ui class

Some of the names and comments are unclear.

Let's refactor them to make it clearer.

yongqqqq committed:
Refactor Parser class

The parse method is handling several operations such as:
* parsing user inputs into arguments
* validating arguments
* creating corresponding command objects

This affects the readability and maintainability of the method.

Let's refactor the method such that:
* separation of parsing logic into separate methods for better clarity.
* creation of helper methods to validate arguments and handle errors.
* separate the command creation logic to improve readability.

yongqqqq committed:
Refactor Storage class

The load method is handling several operations such as:
* reading from file
* file error handling for invalid tasks
* creation of tasks object
* argument validation for each task

This affects the readability and maintainability of the method due
to the multiple responsibilities.

Let's refactor the method such that:
* create separate method for file loading logic from task creation.
* implement separate parsers for each task type.

yongqqqq committed:
Refactor Fishman class

The loadAndSaveTask method is handling several operations such as:
* loading task from storage
* handling different operations
* handling error messages

This affects the readability and maintainability of the method due
to the multiple responsibilities.

Let's refactor the method such that:
* create separate methods to handle each action.
* create separate methods to handle varying line types.

yongqqqq committed:
Update JUnit test

* Refactored existing tests

yongqqqq committed:
Update Javadocs

Changes include:
* Event
* Parser
* Storage
* Fishman

yongqqqq committed:
Add update task functionality

The UpdateCommand feature has been implemented to allow users to update
existing tasks without deleting them.

This will enhance the application by enabling users to update
tasks effectively.

Let's
* implement a new UpdateCommand class to allow users to update
their tasks.
* add error handling to ensure that users key in the correct command.

yongqqqq committed:
Merge pull request #2 from yongqqqq/branch-A-CodeQuality

Increment: A-CodeQuality

yongqqqq committed:
Merge 'master' into 'branch-BCD-CodeQuality'

yongqqqq committed:
Merge pull request #3 from yongqqqq/branch-BCD-Extension

Increment: branch-BCD-Extension

yongqqqq committed:
Fix minor bugs

yongqqqq committed:
Fix checkstyle

yongqqqq committed:
Fix file path issue for dialogbox fxml

yongqqqq committed:
Fix file path issue for dialogbox fxml

yongqqqq committed:
Update Gui for dialog boxes

yongqqqq committed:
Merge branch 'branch-A-BetterGui'

yongqqqq committed:
Update chat bot personality

yongqqqq committed:
Merge remote-tracking branch 'origin/branch-A-Personality'
Merge branch 'branch-A-Personality'

yongqqqq committed:
Update user guide

yongqqqq committed:
Update README.md

yongqqqq committed:
Fix user guide issues

yongqqqq committed:
Fix formatting issue with user guide

yongqqqq committed:
Update README.md

yongqqqq committed:
Update README.md

yongqqqq committed:
Update README.md

yongqqqq committed:
Update README.md

yongqqqq committed:
Update readme

yongqqqq committed:
Update README.md

Add credit for images used

yongqqqq committed:
Add Ui.png
